### PR TITLE
osd: add tier_evict

### PR DESCRIPTION
**Pull request by myoungwon**
add tier_evict() as described in doc/dev/osd_internals/manifest.rst

Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by myoungwon** [comment]
@athanatos Can you take a look?

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -7032,6 +7034,48 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
...
+	}
+	if (!obs.exists) {
+	  result = -ENOENT;
+	  break;
+	}
+	if (get_osdmap()->require_osd_release < ceph_release_t::octopus) {
+	  result = -EOPNOTSUPP;
+	  break;
+	}
+	if (!obs.oi.has_manifest()) {
```
Comment (athanatos) [inline src/osd/PrimaryLogPG.cc:7053]: Should this be an error return?

File: src/osd/PrimaryLogPG.cc [hunk 2]
```diff
@@ -7032,6 +7034,48 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
...
+	if (chunk_length == obs.oi.size) {
+	  // truncate
+	  for (auto p : obs.oi.manifest.chunk_map) {
+	    p.second.set_flag(chunk_info_t::FLAG_MISSING);
+	  }
+	  t->truncate(obs.oi.soid, 0);
+	  ctx->delta_stats.num_bytes -= obs.oi.size;
+	  ctx->delta_stats.num_wr++;
+	  oi.size = 0;
+	  ctx->cache_operation = true;
```
Comment (athanatos) [inline src/osd/PrimaryLogPG.cc:7072]: This works as an initial implementation, though in the future we can hole punch specific chunks if the whole object isn't dedup'd.

**Comment by athanatos** [comment]
This seems about right -- I'll do a real review in the next few days.

**Comment by myoungwon** [comment]
@athanatos Ping. Can you take a look?

**Comment by athanatos** [comment]
This looks ok to me, can you run it through the rados suite?

**Comment by myoungwon** [comment]
retest this please

**Comment by myoungwon** [comment]
@athanatos It seems that none of the failures are related to this change. 

First-run:
https://pulpito.ceph.com/myoungwon-2020-09-30_10:17:32-rados-wip-tier-evict-distro-basic-smithi/

Re-run:
https://pulpito.ceph.com/myoungwon-2020-10-02_01:26:12-rados-wip-tier-evict-distro-basic-smithi/
https://pulpito.ceph.com/myoungwon-2020-10-02_01:20:49-rados-wip-tier-evict-distro-basic-smithi/

**Review by athanatos — APPROVED** [review]

**Pull request merged**